doc/ftprestart.8
.\"
.\" Copyright (c) 1999-2003 WU-FTPD Development Group. 
.\" All rights reserved.
.\" 
.\" Portions Copyright (c) 1980, 1985, 1988, 1989, 1990, 1991, 1993, 1994 
.\" The Regents of the University of California.  Portions Copyright (c) 
.\" 1993, 1994 Washington University in Saint Louis.  Portions Copyright 
.\" (c) 1996, 1998 Berkeley Software Design, Inc.  Portions Copyright (c) 
.\" 1998 Sendmail, Inc.  Portions Copyright (c) 1983, 1995, 1996, 1997 Eric 
.\" P. Allman.  Portions Copyright (c) 1989 Massachusetts Institute of 
.\" Technology.  Portions Copyright (c) 1997 Stan Barber.  Portions 
.\" Copyright (C) 1991, 1992, 1993, 1994, 1995, 1996, 1997 Free Software 
.\" Foundation, Inc.  Portions Copyright (c) 1997 Kent Landfield. 
.\"
.\" Use and distribution of this software and its source code are governed 
.\" by the terms and conditions of the WU-FTPD Software License ("LICENSE"). 
.\"
.\"     $Id: ftprestart.8,v 1.9 2009/05/17 15:18:05 wmaton Exp $
.\"
.TH ftprestart 8
.SH NAME
ftprestart \- restart previously shutdown ftp servers
.SH SYNTAX
.B ftprestart
[
.B \-V
]
.PP
The
.B ftprestart
command
provides an automated way to restart previously shutdown ftp servers.
.B ftprestart
is used when you are ready to re-enable your FTP server. It does the opposite of 
ftpshut and removes shutdown message files that were created by ftpshut. It will 
remove the system-wide shutdown message file as well as the shutdown message files 
in the anonymous ftp areas and any virtual ftp server areas. 
.PP
The
.B \-V
option causes the program to display copyright and version information, then
terminate.

.SH SAMPLE OUTPUT
.PP
  ftprestart: /servers/some.domain/ftp/etc/shutmsg removed. 
  ftprestart: /servers/other.domain/ftp/etc/shutmsg removed. 
  ftprestart: /etc/shutmsg removed. 

.SH SEE ALSO
shutdown(1), ftpshut(8), ftpaccess(5)


CHANGES
 
  Copyright (c) 1999-2003 WU-FTPD Development Group.  
  All rights reserved.
  
  Portions Copyright (c) 1980, 1985, 1988, 1989, 1990, 1991, 1993, 1994
    The Regents of the University of California.
  Portions Copyright (c) 1993, 1994 Washington University in Saint Louis.
  Portions Copyright (c) 1996, 1998 Berkeley Software Design, Inc.
  Portions Copyright (c) 1989 Massachusetts Institute of Technology.
  Portions Copyright (c) 1998 Sendmail, Inc.
  Portions Copyright (c) 1983, 1995, 1996, 1997 Eric P.  Allman.
  Portions Copyright (c) 1997 Stan Barber.
  Portions Copyright (c) 1997 Kent Landfield.
  Portions Copyright (c) 1991, 1992, 1993, 1994, 1995, 1996, 1997
    Free Software Foundation, Inc.  
 
  Use and distribution of this software and its source code are governed 
  by the terms and conditions of the WU-FTPD Software License ("LICENSE").
 
  If you did not receive a copy of the license, it may be obtained online
  at http://www.wu-ftpd.info/license.html.
 
  $Id: CHANGES,v 1.47 2016/03/14 23:57:54 wmaton Exp $

###########################################################################

Changes in 2.8.0-CC5: Released

 o  Updates to config.guess and config.sub to modernise the autoconf
    build system.

 o  configure script regenerated.

 o  Describe auto-buf tuning configure option in README.AUTOCONF.

 o  Add option (-R) to disable REST command in cases where download
    managers can abuse the server by opening multiple connections to
    retrieve chunks of the same file.  Based on a patch by Philippe Troin
    (See Debian bug ID 138497).

 o  Some minor man page cleanups.

 o  More syncs to Debian patches dating back to 2.4.2.

 o  Update INSTALL instructions.

 o  Changed the TLS default cipher from ALL:!EXP to exclude most insecure
    or unsupported ciphers within modern versions of OpenSSL.

Changes in 2.8.0-CC4: Released 7 January, 2012

 o  Thanks to a patch by Laurent CONSTANTIN, calling SSL_library_init()
    will no longer create an error within the SSL code related to a lack
    of ciphers.  The patch is applied to src/tlsutil.c.

 o  Another patch by Laurent CONSTANTIN, this time eliminating extra
    carriage returns that aren't necessary when calling syslog().

 o  Added -c option which displays the compile-time options that were used
    to build WU-FTPD. 

Changes in 2.8.0-CC3: Released 6 November, 2011

 o  Many URL updates.

 o  The default email address used by the server in absence of an email
    directive in ftpaccess is ftp-bugs, and is noted in ftpaccess now.

 o  Added code for ADLER-32 checksums for future extension to the list
    of algorithms available to do checksums.

 o  hard-coded links to cksum and md5sum within the FTP root now point to
    these utilities in /bin instead of /usr/bin.

 o  As defined in GFD.20 (GridFTP:  Protocol Extension to FTP for the
    Grid) by the Global Grid Forum, the SBUF command has been introduced.
    This calls the same code as SITE BUFSIZE would.  SITE BUFSIZE is
    still present for backwards compatibility.

 o  Thanks to a contribution by Austin Ellis, added RHEL6 packaging info
    to the codebase in packages/rpm.  This also includes a patch for
    building on that particular system.

 o  Updated ftpd.8 manpage documenting some missing SITE commands that
    had been in the server since 1998.

 o  First cut at implementing the remainder of the RFC3659 extended FTP
    commands.  FEAT, MLST, OPTS and MLSD are now supported.  Based upon
    early gridFTP releases which used WU-FTPD 2.6.0 and 2.6.1 code.

 o  According to draft-somers-ftp-mfxx-04.txt, defined the stubs for the
    MFF, MFCT and MFMT commands.  These may or may not be supported in the
    future, but at the least they are defined for now.  If so, they may
    become SITE commands instead if the spec is not approved.

 o  Updated doc/examples/ftpconversions based on Debian changes introduced
    sometime in 1995.  The server can now deal with such things as bzip2,
    xz, zip and unzip.

 o  If there's an empty /etc/fstab, a segfault can be produced.  A fix
    from Debian BugID 568235 to path_to_device in src/extensions.c addresses
    this.  Patch by Chris Butler.

 o  Added another possible path for quota.h, patch by Chris Butler of
    Debian.

 o  Added missing handler for SIGTERM, patch by Chris Butler of Debian
    related to Debian bug #536774.

 o  Changed -h invocation on the commandline to print out a very brief help
    listing all commandline options that could be supplied to the daemon.
    Also note that some switches are or have an effect on the configuration
    file.  The former function assigned to the 'h' option is now 'H'.

 o  The former function assigned to the 'h' option is now 'H'.  Said it
    twice now.  The manpage has been updated to reflect the changes.

 o  Implemented the CSID command as found in the IETF draft
    draft-peterson-streamlined-ftp-command-extensions.  This is meant
    to be a standards-based replacement for the CLNT command.  Currently
    this command is hard-wired with the server name and version, but the 
    next release will permit customisation.

###########################################################################

Changes in 2.8.0-CC2: Released 26 June, 2009

 o  Cleaned-up and updated the ftpd.8 manpage which had languished for
    some time.

 o  Added Auto tuning/auto buf, with added commands SITE BUFSIZE and
    SITE BUFSIZEMEASURE.  Use --enable-autobuf to enable this feature.
    Based on a patch on work done on WU-FTPD-2.6.1 by Gaurav Navlakha, et al
    of DAST, NLANR (now CAIDA).  This code also contains the ability to
    report back on buffer usage.  Note that an autobuf client needs to
    be used to take advantage of this feature.  For the time being, the
    code for buffer measurement is IPv4 only.

 o  Updated src/extensions.c to take into account a change in Linux quota
    structure.  Based on a patch by Chris Butler.


src/ftprestart.c
/****************************************************************************  
 
  Copyright (c) 1999-2003 WU-FTPD Development Group.  
  All rights reserved.
  
  Portions Copyright (c) 1980, 1985, 1988, 1989, 1990, 1991, 1993, 1994
    The Regents of the University of California.
  Portions Copyright (c) 1993, 1994 Washington University in Saint Louis.
  Portions Copyright (c) 1996, 1998 Berkeley Software Design, Inc.
  Portions Copyright (c) 1989 Massachusetts Institute of Technology.
  Portions Copyright (c) 1998 Sendmail, Inc.
  Portions Copyright (c) 1983, 1995, 1996, 1997 Eric P.  Allman.
  Portions Copyright (c) 1997 by Stan Barber.
  Portions Copyright (c) 1997 by Kent Landfield.
  Portions Copyright (c) 1991, 1992, 1993, 1994, 1995, 1996, 1997
    Free Software Foundation, Inc.  
 
  Use and distribution of this software and its source code are governed 
  by the terms and conditions of the WU-FTPD Software License ("LICENSE").
 
  If you did not receive a copy of the license, it may be obtained online
  at http://www.wu-ftpd.info/license.html.
 
  $Id: ftprestart.c,v 1.9 2011/10/20 22:58:10 wmaton Exp $
 
****************************************************************************/
/* ftprestart
   **
   ** removes the ftpd shutdown files.
   **
   **  In the previous versions of the wu-ftpd server it was recommended to 
   **  create a link in order for shutdown to work properly for real and 
   **  anonymous user, e.g.  If you use ftpshut, it will create a message 
   **  file at the location specified in the ftpaccess shutdown directive.
   **  ln -s /etc/shutmsg  ~ftp/etc/shutmsg 
   **  
   **  When ftp service is to be restarted after an ftpshut, the shutdown 
   **  message files must be removed. This program reads the ftpaccess
   **  file and finds the location of the system shutdown file.  It
   **  then proceeds to construct a path to the anonymous ftp area with
   **  information found in the "ftp" account.  If virtual ftp servers
   **  are enabled, the shutdown message files within those directories 
   **  are also removed.
   ** 
   **  Initial Author: Kent Landfield
 */
#include "config.h"
#include "proto.h"

#include <errno.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <sys/stat.h>
#include <sys/param.h>
#include <pwd.h>
#if defined(VIRTUAL) && defined(INET6)
#  include <netinet/in.h>
#endif /* defined(VIRTUAL) && defined(INET6) */ 

#include "pathnames.h"

#define MAXVIRTUALS 512

char *progname;
char *msgfiles[MAXVIRTUALS];
int numfiles = 0;

void print_copyright(void);

static int newfile(char *fpath)
{
    int i;
    int fnd;

    /* 
       ** Check to see if the message file path has already been
       ** seen. If so then there is no need to create it again.
     */

    fnd = 0;
    for (i = 0; i < numfiles; i++) {
	if (strcmp(msgfiles[i], fpath) == 0) {
	    fnd = 1;
	    break;
	}
    }
    if (!fnd) {
	msgfiles[numfiles++] = strdup(fpath);
	return (1);
    }
    return (0);
}

static int remove_shutdown_file(char *path)
{
    struct stat stbuf;
    int rc = 1;			/* guilty until proven innocent */

    fprintf(stderr, "%s: %s ", progname, path);

    if (stat(path, &stbuf) == 0) {
	if ((rc = unlink(path)) == 0)
	    fprintf(stderr, "removed.\n");
	else
	    perror(path);
    }
    else
	fprintf(stderr, "does not exist.\n");

    return (rc);
}

int main(int argc, char **argv)
{
    int c;

    char *p;
    char *cp = NULL;
    char linebuf[BUFSIZ];
    char shutmsg[256];
    char anonpath[MAXPATHLEN];
    FILE *accessfile;
    struct passwd *pw;

#if defined(VIRTUAL)
    FILE *svrfp;
    char *sp;
#  if defined(INET6)
    char hostaddress[INET6_ADDRSTRLEN];
#  else /* !(defined(INET6)) */ 
    char hostaddress[32];
#  endif /* !(defined(INET6)) */ 
    char root[MAXPATHLEN];
    char configdir[MAXPATHLEN];
    char accesspath[MAXPATHLEN];
    char altmsgpath[MAXPATHLEN];
    struct stat finfo;
#endif /* defined(VIRTUAL) */ 

    if ((progname = strrchr(argv[0], '/')))
	++progname;
    else
	progname = argv[0];  

    if (argc > 1) {
	while ((c = getopt(argc, argv, "V")) != EOF) {
	    switch (c) {
	    case 'V':
		print_copyright();
		exit(0);
	    default:
		fprintf(stderr, "usage: %s [-V]\n", progname);
		exit(1);
	    }
	}
    }

    if ((accessfile = fopen(_PATH_FTPACCESS, "r")) == NULL) {
	if (errno != ENOENT)
	    fprintf(stderr, "%s: could not open access file %s: %s\n",
		    progname, _PATH_FTPACCESS, strerror(errno));
	exit(1);
    }

    /* 
       ** Search the access file for the 'shutdown' directive.
     */

    while (fgets(linebuf, BUFSIZ, accessfile) != NULL) {
	if (strncasecmp(linebuf, "shutdown", 8) == 0) {
	    (void) strtok(linebuf, " \t");
	    strlcpy(shutmsg, strtok(NULL, " \t"), sizeof(shutmsg));
	    cp = shutmsg;
	    if ((p = strchr(cp, '\n')) != NULL)
		*p = '\0';
	}
    }

    if (cp == NULL) {
	fprintf(stderr, "%s: no shutdown path defined in ftpaccess file %s.\n",
                progname, _PATH_FTPACCESS);
	exit(1);
    }

    msgfiles[numfiles++] = shutmsg;

    /*
       ** Get the location of the anonymous ftp area and check
       ** to see if there is a file shutdown file there as well. 
       ** If so, remove it.
     */
    if ((pw = getpwnam("ftp")) != NULL) {
	snprintf(anonpath, sizeof(anonpath), "%s%s", pw->pw_dir, shutmsg);
	if (newfile(anonpath))
	    (void) remove_shutdown_file(anonpath);
    }

#if defined(VIRTUAL)
    /*
       ** Search the access file for virtual ftp servers.
       ** If found, check if there are links/shutdown
       ** message files files in the virtual server areas.
       ** If so, remove them.
     */

    rewind(accessfile);

    while (fgets(linebuf, sizeof(linebuf) - 1, accessfile) != NULL) {
	if (strncasecmp(linebuf, "virtual", 7) == 0) {
	    if ((p = strstr(linebuf, "root")) != NULL) {
		p += 4;

		if ((cp = strchr(linebuf, '\n')) != NULL)
		    *cp = '\0';

		/* skip to the path */

		while (*p && isspace(*p))
		    p++;
		cp = p;
		while (*p && isalnum(*p))
		    p++;

		snprintf(altmsgpath, sizeof(altmsgpath), "%s%s", cp, shutmsg);
		if (newfile(altmsgpath))
		    (void) remove_shutdown_file(altmsgpath);
	    }
	}
    }


    /*
       ** Need to deal with the access files at the virtual domain directory
       ** locations specified in the ftpservers file.
     */

    if ((svrfp = fopen(_PATH_FTPSERVERS, "r")) != NULL) {
	while (read_servers_line(svrfp, hostaddress, sizeof(hostaddress), configdir, sizeof(configdir)) == 1) {
	    /* get rid of any trailing slash */
	    sp = configdir + (strlen(configdir) - 1);
	    if (*sp == '/')
		*sp = '\0';

	    /*
	       ** check to see that a valid directory value was
	       ** supplied and not something such as "INTERNAL"
	       **
	       ** It is valid to have a string such as "INTERNAL" in the
	       ** ftpservers entry. This is not an error. Silently ignore it.
	     */

	    if ((stat(configdir, &finfo) < 0) ||
		((finfo.st_mode & S_IFMT) != S_IFDIR))
		continue;

	    snprintf(accesspath, sizeof(accesspath), "%s/ftpaccess", configdir);

	    (void) fclose(accessfile);

	    if ((accessfile = fopen(accesspath, "r")) == NULL) {
		if (errno != ENOENT) {
		    fprintf(stderr, "%s: could not open access file %s: %s\n",
			    progname, accesspath, strerror(errno));
		    continue;
		}
	    }

	    /* need to find the root path */

	    while (fgets(linebuf, sizeof(linebuf) - 1, accessfile) != NULL) {
		if ((sp = strstr(linebuf, "root")) != NULL) {
		    if ((cp = strchr(sp, '\n')) != NULL)
			*cp = '\0';	/* strip newline */
		    sp += 4;	/* skip past "root" keyword */

		    while (*sp && isspace(*sp))		/* skip whitespace to path */
			sp++;
		    cp = sp;
		    while (*sp && !isspace(*sp))
			sp++;
		    *sp = '\0';	/* truncate blanks, comments etc. */
		    strlcpy(root, cp, sizeof(root));
		    break;
		}
	    }

	    rewind(accessfile);

	    /* need to find the shutdown message file path */

	    while (fgets(linebuf, sizeof(linebuf) - 1, accessfile) != NULL) {
		if ((sp = strstr(linebuf, "shutdown")) != NULL) {
		    if ((cp = strchr(sp, '\n')) != NULL)
			*cp = '\0';	/* strip newline */
		    sp += 8;	/* skip past "root" keyword */

		    while (*sp && isspace(*sp))	/* skip whitespace to path */
			sp++;
		    cp = sp;
		    while (*sp && !isspace(*sp))
			sp++;
		    *sp = '\0';	/* truncate blanks, comments etc. */
		    break;
		}
	    }

	    /*
	       ** check to make sure the admin hasn't specified 
	       ** a complete path in the 'shutdown' directive.
	     */
	    if ((sp = strstr(cp, root)) == NULL)
		snprintf(altmsgpath, sizeof(altmsgpath), "%s%s", root, cp);

	    if (newfile(altmsgpath))
		(void) remove_shutdown_file(altmsgpath);
	}
	fclose(svrfp);
    }
#endif /* defined(VIRTUAL) */ 

    fclose(accessfile);

    /*
       ** Time to remove the system wide shutdown file.
     */
    return (remove_shutdown_file(shutmsg));
}
